File: dsa-problems/Strings/solutions/cpp/STR-003-smallest-missing-substring.cpp
#include <iostream>
class Solution {
private:
    string dfs(string current, int remaining, unordered_set<string>& substrings) {
        if (remaining == 0) {
            return substrings.count(current) ? "" : current;
        }

        for (char c = 'a'; c <= 'z'; c++) {
            current += c;
            string result = dfs(current, remaining - 1, substrings);
            if (!result.empty()) {
                return result;
            }
            current.pop_back(); // backtrack
        }

        return "";
    }

public:
    string smallestMissingSubstring(string s, int k) {
        // Extract all k-length substrings
        unordered_set<string> substrings;
        for (int i = 0; i <= (int)s.size() - k; i++) {
            substrings.insert(s.substr(i, k));
        }

        return dfs("", k, substrings);
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr);
    string s; cin >> s;
    int k; cin >> k;
    Solution sol;
    cout << sol.smallestMissingSubstring(s, k) << endl;
    return 0;
}

File: dsa-problems/Strings/solutions/java/STR-003-smallest-missing-substring.java
class Solution {
    public String smallestMissingSubstring(String s, int k) {
        // Extract all k-length substrings
        Set<String> substrings = new HashSet<>();
        for (int i = 0; i <= s.length() - k; i++) {
            substrings.add(s.substring(i, i + k));
        }

        // DFS to find smallest missing
        return dfs(new StringBuilder(), k, substrings);
    }

    private String dfs(StringBuilder current, int remaining, Set<String> substrings) {
        if (remaining == 0) {
            String candidate = current.toString();
            return substrings.contains(candidate) ? null : candidate;
        }

        for (char c = 'a'; c <= 'z'; c++) {
            current.append(c);
            String result = dfs(current, remaining - 1, substrings);
            if (result != null) {
                return result;
            }
            current.setLength(current.length() - 1); // backtrack
        }

        return null;
    }
}

class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.next();
        int k = sc.nextInt();
        Solution sol = new Solution();
        System.out.println(sol.smallestMissingSubstring(s, k));
        sc.close();
    }
}
